### PR TITLE
Create a Petal "Edit this Page" button for HTML pages

### PR DESCRIPTION
**Pull request by plutonik-a**
### PR Description

This PR will create a HTML section containing a link button with a redirect `href` attribute to the Petal web application.

* set default values for mvn parameters, which can be overridden with custom properties when calling the build script for generating the website
* update build script for GitHub Action with custom properties for generating the website
* create a "Petal Edit this Page" for each topic page that contains a `href` attribute with required parameters for a valid redirect to the Petal Editor Application
* Parameters for the button contain:
    - `ghrepo` (`$petal-github-base-url` + `$petal-github-org-name` + `$petal-github-repo-name`)
    - `source` (xslt string processing),
    - `branch` (`$petal-github-branch`)
    - `referer` (`$petal-referrer-url`)

#### General Test

1. Generate the website with the default parameters: `mvn clean package -Pquick-start-guide-website`.
2. Open any of the generated HTML files in path `cityehr-quick-start-guide/target/website`.
3. Check the href attribute of the button under the top navigation with label "Edit this Page". It should contain the above listed parameters with the correct values.
4. Click on the link to see at least the correct redirect to the Petal production site. 

#### Test the GitHub Integration

1. Generate the website with a custom `petal.base-url` parameter to point to the Petal server instance with the running GitHub integration feature: E.g. `mvn clean -Dpetal.base-url=<YOUR_CUSTOM_URL_HERE> package -Pquick-start-guide-website`

### CODE REVIEW

File: cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt [hunk 1]
```diff
@@ -25,15 +34,29 @@
         </xsl:call-template>
       </head>
       <body>
-        <div id="top-nav">
+        <nav id="top-nav">
```
Comment (adamretter) [inline cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt:36]: Why is this PR changing something unrelated to its purpose?

File: cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt [hunk 2]
```diff
@@ -25,15 +34,29 @@
...
+          <a href="{ $petal-full-url }">
+            <input type="button" value="Edit this page" />
+          </a>
+        </section>
+
         <article>
           <xsl:apply-templates select="element()" mode="body"/>
         </article>
-        <div id="bottom-nav">
+        <nav id="bottom-nav">
```
Comment (adamretter) [inline cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt:54]: Why is this PR changing something unrelated to its purpose?

File: cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt [hunk 3]
```diff
@@ -25,15 +34,29 @@
...
       </head>
       <body>
-        <div id="top-nav">
+        <nav id="top-nav">
           <xsl:apply-templates select="." mode="top-nav"/>
-        </div>
+        </nav>
+
+        <!-- Petal Edit Button -->
+        <section>
```
Comment (adamretter) [inline cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt]: As I understand the use of `section` tag in HTML 5, it shouldn't really be used for this purpose. Should this go in the `nav`, or perhaps a separate `div` with id would be more appropriate?

File: cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt [hunk 4]
```diff
@@ -25,15 +34,29 @@
...
+          <xsl:variable name="petal-source-file-uri" select="com:document-uri(.)" />
+          <xsl:variable name="tokens" select="tokenize($petal-source-file-uri, '/')"/>
+          <xsl:variable name="orig-file-name" select="$tokens[last()]"/>
+          <xsl:variable name="petal-source-file" select="substring-after($petal-source-file-uri, $petal-github-repo-name || '/')" />
```
Comment (adamretter) [inline cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt:43]: We already have a function for obtaining the filename, you could replace:

```xslt
<xsl:variable name="petal-source-file-uri" select="com:document-uri(.)" />
<xsl:variable name="tokens" select="tokenize($petal-source-file-uri, '/')"/>
<xsl:variable name="orig-file-name" select="$tokens[last()]"/>
<xsl:variable name="petal-source-file" select="substring-after($petal-source-file-uri, $petal-github-repo-name || '/')" />
<xsl:variable name="petal-file-name" select="hcom:dita-filename-to-html($orig-file-name)" />
```

with just:
```xslt
<xsl:variable name="petal-source-file" select="com:filename(com:document-uri(.))" />
<xsl:variable name="petal-file-name" select="hcom:dita-filename-to-html($petal-source-file)" />
```

Could you also rename `petal-file-name` to something more descriptive please? Perhaps `petal-webpage-filename`.

File: cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt [hunk 5]
```diff
@@ -25,15 +34,29 @@
...
+          <xsl:variable name="petal-file-name" select="hcom:dita-filename-to-html($orig-file-name)" />
+          <xsl:variable name="petal-full-url" select="concat($petal-base-url, '?ghrepo=', $petal-github-base-url, '/', $petal-github-org-name, '/', $petal-github-repo-name, '&amp;source=', $petal-source-file, '&amp;branch=', $petal-github-branch, '&amp;referer=', $petal-referrer-url, $petal-file-name)" />
```
Comment (adamretter) [inline cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt]: I am not sure that we have the value for `ghrepo` correct? At the moment it will produce:

```
?ghrepo=https://github.com/evolvedbinary/cityehr-documentation
```

The `https://github.com/` part seem superfluous as this is fixed and won't change. I think we should just be able to send:
```
?ghrepo=evolvedbinary/cityehr-documentation
```

We could then remove the `petal-github-base-url` from the XSLT, Maven, and CI scripts.

File: cityehr-quick-start-guide/pom.xml [hunk 6]
```diff
@@ -25,6 +25,13 @@
         <fo.generated-resources>${project.build.directory}/generated-resources/fo</fo.generated-resources>
         <html.generated-resources>${project.build.directory}/generated-resources/html</html.generated-resources>
         <website.output.folder>${project.build.directory}/website</website.output.folder>
+
+        <petal.base-url>https://petal.evolvedbinary.com/</petal.base-url>
```
Comment (adamretter) [inline cityehr-quick-start-guide/pom.xml]: Could we rename `petal-base-url` to `petal-api-url` please? (also in XSLT, and CI scripts)

File: cityehr-quick-start-guide/pom.xml [hunk 7]
```diff
@@ -25,6 +25,13 @@
...
+        <petal.github-base-url>https://github.com/</petal.github-base-url>
+        <petal.github-org-name>evolvedbinary</petal.github-org-name>
+        <petal.github-repo-name>cityehr-documentation</petal.github-repo-name>
+        <petal.github-branch>develop</petal.github-branch>
+        <petal.referrer-url>https://evolvedbinary.github.io/cityehr-documentation/</petal.referrer-url>
```
Comment (adamretter) [inline cityehr-quick-start-guide/pom.xml]: Can you rename `petal-referer-url` to `petal.referrer-base-url` please (also in XSLT and CI scripts), and drop the trailing `/` from here (the add it in the concat inside the XSLT instead).

File: cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt [hunk 8]
```diff
@@ -25,15 +34,29 @@
...
+          <a href="{ $petal-full-url }">
```
Comment (adamretter) [inline cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt]: `href="{ $petal-full-url }"` should not have the extra whitespace please, e.g. change to `href="{$petal-full-url}"`

**Review by adamretter — CHANGES_REQUESTED** [review]
In general this looks good, just some small suggestions on improving it now.

File: cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt (hunk 1 above)
Comment (plutonik-a) [reply cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt:36]: @adamretter I hoped you wouldn't mind about this small improvement towards semantical markup, since I was working on the same part of HTML.

File: cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt (hunk 2 above)
Comment (plutonik-a) [reply cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt:54]: See above.

File: cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt (hunk 3 above)
Comment (plutonik-a) [reply cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt]: @adamretter I see your point, and the `section` element is not the best choice here. 
The button is not part of a navigation. The `nav` is already represented by your navigation element on top and at the end of the page.
The best practice for choosing a `div` is to use it only for layout reasons, because it has no semantical meaning.
Therefore I went for the `section`, even though it is a stand-alone button at the moment, but I hoped we could add some information next to it, perhaps?
I could use an `aside` element, that represents supplementary content which is not part of the main document. How about that?

File: cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt (hunk 5 above)
Comment (plutonik-a) [reply cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt]: @adamretter Fixed with f8a9252 and 1e0be49

File: cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt (hunk 4 above)
Comment (plutonik-a) [reply cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt:43]: @adamretter Fixed the renaming of the variable `petal-file-name` to `petal-webpage-filename` with 93e0906.

Regarding the usage of function `hcom:dita-filename-to-html()`:
I am already using it, but the difficulty is to retrieve and pass the correct argument to the function. 
Hence the complex string processing   to get the correct value for `$petal-source-file` that you use in your example above.
 
Btw, your suggestion would return a filename like this `cityehr-quick-start-guide/src/main/lwdita/quickstart-guide-modular/introduction.html`, but we need the `.xdita` file ending (the GitHub URI), therefore my initial processing of the document URI. 

To summarize the required URL parts of example "introduction.html":
`$petal-base-url`  -> `https://petal.evolvedbinary.com/`
`?ghrepo=`  -> `evolvedbinary/cityehr-documentation`
`&source=` -> `cityehr-quick-start-guide/src/main/lwdita/quickstart-guide-modular/introduction.dita`
`&branch=` -> `develop`
`&referer=` -> `https://evolvedbinary.github.io/cityehr-documentation/introduction.html`

File: cityehr-quick-start-guide/pom.xml (hunk 6 above)
Comment (plutonik-a) [reply cityehr-quick-start-guide/pom.xml]: Fixed with a12f8e5

File: cityehr-quick-start-guide/pom.xml (hunk 7 above)
Comment (plutonik-a) [reply cityehr-quick-start-guide/pom.xml]: Fixed with ad76cc3

File: cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt (hunk 8 above)
Comment (plutonik-a) [reply cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt]: Fixed with 135a400

File: cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt (hunk 1 above)
Comment (adamretter) [reply cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt:36]: It should really be in separate PR as it is not related to the changes in this PR. Next time please @plutonik-a.

File: cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt (hunk 3 above)
Comment (adamretter) [reply cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt]: @plutonik-a As I understand the use of section tag in HTML 5, an `aside` isn't suitable here. Please change this to a `div` with an appropriate `id`.

Comment (plutonik-a) [reply cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt]: Fixed with a3b2751

File: cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt (hunk 4 above)
Comment (adamretter) [reply cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt:43]: @alex how about just:

```xslt
<xsl:variable name="petal-source-file-uri" select="com:document-uri(.)" />
<xsl:variable name="petal-source-file" select="substring-after($petal-source-file-uri, $petal-github-repo-name || '/')" />
<xsl:variable name="petal-webpage-filename" select="hcom:dita-filename-to-html(com:filename($petal-source-file-uri))" />
<xsl:variable name="petal-full-url" select="concat($petal-api-url, '?ghrepo=', $petal-github-org-name, '/', $petal-github-repo-name, '&amp;source=', $petal-source-file, '&amp;branch=', $petal-github-branch, '&amp;referer=', $petal-referrer-base-url, $petal-webpage-filename)" />
<a href="{$petal-full-url}">
    <input type="button" value="Edit this page" />
</a>
```

File: .github/workflows/pages.yml [hunk 9]
```diff
@@ -30,7 +30,8 @@ jobs:
           distribution: 'liberica'
       - name: Generate website (XSLT)
         id: generate
-        run: mvn -B package -Pquick-start-guide-website
+        # Override Maven properties for the Petal button url
```
Comment (adamretter) [inline .github/workflows/pages.yml:33]: @plutonik-a Can you remove the trailing `/` 's please? They should be computed in the XSLT.

File: cityehr-quick-start-guide/pom.xml [hunk 10]
```diff
@@ -25,6 +25,12 @@
         <fo.generated-resources>${project.build.directory}/generated-resources/fo</fo.generated-resources>
         <html.generated-resources>${project.build.directory}/generated-resources/html</html.generated-resources>
         <website.output.folder>${project.build.directory}/website</website.output.folder>
+
```
Comment (adamretter) [inline cityehr-quick-start-guide/pom.xml:28]: @plutonik-a Can you remove the trailing `/` 's please? They should be computed in the XSLT.

File: cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt (hunk 4 above)
Comment (plutonik-a) [reply cityehr-quick-start-guide/src/main/xslt/create-topic-html.xslt:43]: Yes, thanks for the hint, that fixes it! Fixed with b3f5cbc.

File: .github/workflows/pages.yml (hunk 9 above)
Comment (plutonik-a) [reply .github/workflows/pages.yml:33]: Fixed with 5517d0c

File: cityehr-quick-start-guide/pom.xml (hunk 10 above)
Comment (plutonik-a) [reply cityehr-quick-start-guide/pom.xml:28]: Fixed with https://github.com/evolvedbinary/cityehr-documentation/commit/5517d0c310ce2d9abfc5caa9155c441f23ac4d19
